From ba249850fc6d8225b063c76b90a7bf6f9d3592ff Mon Sep 17 00:00:00 2001
From: Taozuhong <taozuhong@hotmail.com>
Date: Mon, 12 Oct 2026 00:14:33 +0000
Subject: [PATCH] MinGW: port build system and restore WTF Windows headers

Build-system port for the GTK/MinGW target:
- gio-unix-2.0 -> gio-2.0, WIN32 option defaults, mingw-compat include
  dir wiring, WTF CMakeLists adjustments.
- Restore 4 WTF Windows headers that are stripped from the release
  tarball (text/win/WCharStringExtras.h, win/SoftLinking.h,
  win/DbgHelperWin.h, win/WTFCRTDebug.h) and register them in the
  WTF_PUBLIC_HEADERS WIN32 list; WTFCRTDebug.h provides an inline
  no-op disableCRTDebugAssertDialog() under MinGW (its .cpp is not
  in the tarball while jsc.cpp calls it).
---
 Source/WTF/wtf/CMakeLists.txt               |  10 +-
 Source/WTF/wtf/text/win/WCharStringExtras.h |  60 ++++
 Source/WTF/wtf/win/DbgHelperWin.h           |  41 +++
 Source/WTF/wtf/win/SoftLinking.h            | 313 ++++++++++++++++++++
 Source/WTF/wtf/win/WTFCRTDebug.h            |  40 +++
 Source/cmake/FindGLib.cmake                 |   8 +-
 Source/cmake/OptionsGTK.cmake               |  28 ++
 7 files changed, 497 insertions(+), 3 deletions(-)
 create mode 100644 Source/WTF/wtf/text/win/WCharStringExtras.h
 create mode 100644 Source/WTF/wtf/win/DbgHelperWin.h
 create mode 100644 Source/WTF/wtf/win/SoftLinking.h
 create mode 100644 Source/WTF/wtf/win/WTFCRTDebug.h

diff --git a/Source/WTF/wtf/CMakeLists.txt b/Source/WTF/wtf/CMakeLists.txt
index fda02527..ea4640a8 100644
--- a/Source/WTF/wtf/CMakeLists.txt
+++ b/Source/WTF/wtf/CMakeLists.txt
@@ -708,13 +708,19 @@ set(WTF_SOURCES
 )
 
 if (WIN32)
+    # The 2.54.1 release tarball strips most wtf/win/ and wtf/text/win/
+    # headers (unused by the Unix ports), yet WTFString.h, SoftLinking.h,
+    # StackTrace.h and jsc.cpp still include them under OS(WINDOWS). The
+    # tarball's header list kept only win/Win32Handle.h, so MinGW builds
+    # failed to resolve the rest; the missing headers are restored by the
+    # MinGW port patches and are registered here so they are copied into
+    # ${WTF_HEADERS_DIR} together with the rest of the public headers.
     list(APPEND WTF_PUBLIC_HEADERS
         text/win/WCharStringExtras.h
 
         win/DbgHelperWin.h
-        win/GDIObject.h
-        win/PathWalker.h
         win/SoftLinking.h
+        win/WTFCRTDebug.h
         win/Win32Handle.h
     )
 elseif (APPLE)
diff --git a/Source/WTF/wtf/text/win/WCharStringExtras.h b/Source/WTF/wtf/text/win/WCharStringExtras.h
new file mode 100644
index 00000000..3412ae20
--- /dev/null
+++ b/Source/WTF/wtf/text/win/WCharStringExtras.h
@@ -0,0 +1,60 @@
+/*
+ * Copyright (C) 2017 Konstantin Tokarev <annulen@yandex.ru>
+ * Copyright (C) 2019 Sony Interactive Entertainment Inc.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <unicode/utypes.h>
+
+namespace WTF {
+
+inline const char16_t* ucharFrom(const wchar_t* characters)
+{
+    static_assert(sizeof(wchar_t) == sizeof(char16_t), "We assume wchar_t and char16_t have the same size");
+    return reinterpret_cast<const char16_t*>(characters);
+}
+
+inline char16_t* ucharFrom(wchar_t* characters)
+{
+    static_assert(sizeof(wchar_t) == sizeof(char16_t), "We assume wchar_t and char16_t have the same size");
+    return reinterpret_cast<char16_t*>(characters);
+}
+
+inline const wchar_t* wcharFrom(const char16_t* characters)
+{
+    static_assert(sizeof(wchar_t) == sizeof(char16_t), "We assume wchar_t and char16_t have the same size");
+    return reinterpret_cast<const wchar_t*>(characters);
+}
+
+inline wchar_t* wcharFrom(char16_t* characters)
+{
+    static_assert(sizeof(wchar_t) == sizeof(char16_t), "We assume wchar_t and char16_t have the same size");
+    return reinterpret_cast<wchar_t*>(characters);
+}
+
+} // namespace WTF
+
+using WTF::ucharFrom;
+using WTF::wcharFrom;
diff --git a/Source/WTF/wtf/win/DbgHelperWin.h b/Source/WTF/wtf/win/DbgHelperWin.h
new file mode 100644
index 00000000..d951a0b0
--- /dev/null
+++ b/Source/WTF/wtf/win/DbgHelperWin.h
@@ -0,0 +1,41 @@
+/*
+ * Copyright (C) 2018 Sony Interactive Entertainment Inc.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE AND ITS CONTRIBUTORS "AS IS" AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL APPLE OR ITS CONTRIBUTORS BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <windows.h>
+#include <wtf/Lock.h>
+#include <wtf/Noncopyable.h>
+
+#include <dbghelp.h>
+
+namespace WTF {
+
+namespace DbgHelper {
+
+WTF_EXPORT_PRIVATE bool SymFromAddress(HANDLE hProc, DWORD64 address, DWORD64* displacement, SYMBOL_INFO*);
+
+};
+
+} // namespace WTF
diff --git a/Source/WTF/wtf/win/SoftLinking.h b/Source/WTF/wtf/win/SoftLinking.h
new file mode 100644
index 00000000..dd4a64ce
--- /dev/null
+++ b/Source/WTF/wtf/win/SoftLinking.h
@@ -0,0 +1,313 @@
+/*
+ * Copyright (C) 2007, 2009-2014 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1.  Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ * 2.  Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <windows.h>
+#include <wtf/Assertions.h>
+
+#pragma mark - Soft-link helper macros
+
+#define SOFT_LINK_LIBRARY_HELPER(lib, suffix) \
+    static HMODULE lib##Library() \
+    { \
+        static HMODULE library = LoadLibraryW(L###lib suffix); \
+        return library; \
+    }
+
+#define SOFT_LINK_GETPROCADDRESS GetProcAddress
+#define SOFT_LINK_LIBRARY(lib) SOFT_LINK_LIBRARY_HELPER(lib, L".dll")
+#define SOFT_LINK_DEBUG_LIBRARY(lib) SOFT_LINK_LIBRARY_HELPER(lib, L"_debug.dll")
+
+#pragma mark - Soft-link macros for use within a single source file
+
+#define SOFT_LINK(library, functionName, resultType, callingConvention, parameterDeclarations, parameterNames) \
+    static void* softLink##functionName; \
+    \
+    inline resultType functionName parameterDeclarations \
+    { \
+        if (!softLink##functionName) \
+            softLink##functionName = ::EncodePointer(reinterpret_cast<void*>(SOFT_LINK_GETPROCADDRESS(library##Library(), #functionName))); \
+        return reinterpret_cast<resultType (callingConvention*) parameterDeclarations>(::DecodePointer(softLink##functionName)) parameterNames; \
+    }
+
+#define SOFT_LINK_OPTIONAL(library, functionName, resultType, callingConvention, parameterDeclarations) \
+    typedef resultType (callingConvention *functionName##PtrType) parameterDeclarations; \
+    static functionName##PtrType functionName##Ptr() \
+    { \
+        static void* ptr; \
+        static bool initialized; \
+        \
+        if (!initialized) { \
+            ptr = ::EncodePointer(reinterpret_cast<void*>(SOFT_LINK_GETPROCADDRESS(library##Library(), #functionName))); \
+            initialized = true; \
+        } \
+        return reinterpret_cast<functionName##PtrType>(::DecodePointer(ptr)); \
+    } \
+
+#define SOFT_LINK_LOADED_LIBRARY(library, functionName, resultType, callingConvention, parameterDeclarations) \
+    typedef resultType (callingConvention *functionName##PtrType) parameterDeclarations; \
+    static functionName##PtrType functionName##Ptr() \
+    { \
+        static void* ptr; \
+        static bool initialized; \
+        \
+        if (!initialized) { \
+            static HINSTANCE libraryInstance = ::GetModuleHandle(L#library); \
+            ptr = ::EncodePointer(reinterpret_cast<void*>(SOFT_LINK_GETPROCADDRESS(libraryInstance, #functionName))); \
+            initialized = true; \
+        } \
+        \
+        return reinterpret_cast<functionName##PtrType>(::DecodePointer(ptr)); \
+    } \
+
+/*
+    In order to soft link against functions decorated with __declspec(dllimport), we prepend "softLink_" to the function names.
+    If you use SOFT_LINK_DLL_IMPORT(), you will also need to #define the function name to account for this, e.g.:
+
+    SOFT_LINK_DLL_IMPORT(myLibrary, myFunction, ...)
+    #define myFunction softLink_myFunction
+*/
+#define SOFT_LINK_DLL_IMPORT(library, functionName, resultType, callingConvention, parameterDeclarations, parameterNames) \
+    static void* softLink##functionName; \
+    \
+    inline resultType softLink_##functionName parameterDeclarations \
+    { \
+        if (!softLink##functionName) \
+            softLink##functionName = ::EncodePointer(reinterpret_cast<void*>(SOFT_LINK_GETPROCADDRESS(library##Library(), #functionName))); \
+        return reinterpret_cast<resultType(callingConvention*)parameterDeclarations>(::DecodePointer(softLink##functionName)) parameterNames; \
+    }
+
+#define SOFT_LINK_DLL_IMPORT_OPTIONAL(library, functionName, resultType, callingConvention, parameterDeclarations) \
+    typedef resultType (callingConvention *functionName##PtrType) parameterDeclarations; \
+    static functionName##PtrType functionName##Ptr() \
+    { \
+        static void* ptr; \
+        static bool initialized; \
+        \
+        if (!initialized) { \
+            ptr = ::EncodePointer(reinterpret_cast<void*>(SOFT_LINK_GETPROCADDRESS(library##Library(), #functionName))); \
+            initialized = true; \
+        } \
+        return reinterpret_cast<functionName##PtrType>(::DecodePointer(ptr)); \
+    } \
+
+#define SOFT_LINK_DLL_IMPORT_OPTIONAL(library, functionName, resultType, callingConvention, parameterDeclarations) \
+    typedef resultType (callingConvention *functionName##PtrType) parameterDeclarations; \
+    static functionName##PtrType functionName##Ptr() \
+    { \
+        static void* ptr; \
+        static bool initialized; \
+        \
+        if (!initialized) { \
+            ptr = ::EncodePointer(reinterpret_cast<void*>(SOFT_LINK_GETPROCADDRESS(library##Library(), #functionName))); \
+            initialized = true; \
+        } \
+        return reinterpret_cast<functionName##PtrType>(::DecodePointer(ptr)); \
+    } \
+
+/*
+    Variables exported by a DLL need to be accessed through a function.
+    If you use SOFT_LINK_VARIABLE_DLL_IMPORT(), you will also need to #define the variable name to account for this, e.g.:
+
+    SOFT_LINK_VARIABLE_DLL_IMPORT(myLibrary, myVar, int)
+    #define myVar get_myVar()
+*/
+#define SOFT_LINK_VARIABLE_DLL_IMPORT(library, variableName, variableType) \
+    static variableType get##variableName() \
+    { \
+        static variableType* ptr = reinterpret_cast<variableType*>(SOFT_LINK_GETPROCADDRESS(library##Library(), #variableName)); \
+        ASSERT(ptr); \
+        return *ptr; \
+    } \
+
+/*
+    Note that this will only work for variable types for which a return value of 0 can signal an error.
+ */
+#define SOFT_LINK_VARIABLE_DLL_IMPORT_OPTIONAL(library, variableName, variableType) \
+    static variableType get##variableName() \
+    { \
+        static variableType* ptr = reinterpret_cast<variableType*>(SOFT_LINK_GETPROCADDRESS(library##Library(), #variableName)); \
+        if (!ptr) \
+            return 0; \
+        return *ptr; \
+    } \
+
+#pragma mark - Soft-link macros for sharing across multiple source files
+
+// See Source/WebCore/platform/cf/CoreMediaSoftLink.{cpp,h} for an example implementation.
+
+#define SOFT_LINK_FRAMEWORK_FOR_HEADER(functionNamespace, framework) \
+    namespace functionNamespace { \
+    extern HMODULE framework##Library(bool isOptional = false); \
+    bool is##framework##FrameworkAvailable(); \
+    inline bool is##framework##FrameworkAvailable() { \
+        return framework##Library(true) != nullptr; \
+    } \
+    }
+
+#define SOFT_LINK_FRAMEWORK_HELPER(functionNamespace, framework, suffix, export) \
+    namespace functionNamespace { \
+    export HMODULE framework##Library(bool isOptional = false); \
+    export HMODULE framework##Library(bool isOptional) \
+    { \
+        static HMODULE library = LoadLibraryW(L###framework suffix); \
+        ASSERT_WITH_MESSAGE_UNUSED(isOptional, isOptional || library, "Could not load %s", L###framework suffix); \
+        return library; \
+    } \
+    }
+
+#define SOFT_LINK_FRAMEWORK(functionNamespace, framework) SOFT_LINK_FRAMEWORK_HELPER(functionNamespace, framework, L".dll", )
+#define SOFT_LINK_DEBUG_FRAMEWORK(functionNamespace, framework) SOFT_LINK_FRAMEWORK_HELPER(functionNamespace, framework, L"_debug.dll", )
+
+#ifdef DEBUG_ALL
+#define SOFT_LINK_FRAMEWORK_FOR_SOURCE(functionNamespace, framework) SOFT_LINK_DEBUG_FRAMEWORK(functionNamespace, framework)
+#define SOFT_LINK_FRAMEWORK_FOR_SOURCE_WITH_EXPORT(functionNamespace, framework, export) SOFT_LINK_FRAMEWORK_HELPER(functionNamespace, framework, L"_debug.dll", export)
+#else
+#define SOFT_LINK_FRAMEWORK_FOR_SOURCE(functionNamespace, framework) SOFT_LINK_FRAMEWORK(functionNamespace, framework)
+#define SOFT_LINK_FRAMEWORK_FOR_SOURCE_WITH_EXPORT(functionNamespace, framework, export) SOFT_LINK_FRAMEWORK_HELPER(functionNamespace, framework, L".dll", export)
+#endif
+
+
+
+#define SOFT_LINK_CONSTANT_FOR_HEADER(functionNamespace, framework, variableName, variableType) \
+    namespace functionNamespace { \
+    variableType get_##framework##_##variableName(); \
+    }
+
+#define SOFT_LINK_CONSTANT_FOR_SOURCE_WITH_EXPORT(functionNamespace, framework, variableName, variableType, export) \
+    namespace functionNamespace { \
+    static void init##framework##variableName(void* context) { \
+        variableType* ptr = reinterpret_cast<variableType*>(SOFT_LINK_GETPROCADDRESS(framework##Library(), #variableName)); \
+        RELEASE_ASSERT(ptr); \
+        *static_cast<variableType*>(context) = *ptr; \
+    } \
+    export variableType get_##framework##_##variableName(); \
+    variableType get_##framework##_##variableName() \
+    { \
+        static variableType constant##framework##variableName; \
+        static dispatch_once_t once; \
+        dispatch_once_f(&once, static_cast<void*>(&constant##framework##variableName), init##framework##variableName); \
+        return constant##framework##variableName; \
+    } \
+    }
+
+#define SOFT_LINK_CONSTANT_FOR_SOURCE(functionNamespace, framework, variableName, variableType) \
+    SOFT_LINK_CONSTANT_FOR_SOURCE_WITH_EXPORT(functionNamespace, framework, variableName, variableType, )
+
+#define SOFT_LINK_CONSTANT_MAY_FAIL_FOR_HEADER(functionNamespace, framework, variableName, variableType) \
+    namespace functionNamespace { \
+    bool canLoad_##framework##_##variableName(); \
+    bool init_##framework##_##variableName(); \
+    variableType get_##framework##_##variableName(); \
+    }
+
+#define SOFT_LINK_CONSTANT_MAY_FAIL_FOR_SOURCE(functionNamespace, framework, variableName, variableType) \
+    namespace functionNamespace { \
+    static variableType constant##framework##variableName; \
+    bool init_##framework##_##variableName(); \
+    bool init_##framework##_##variableName() \
+    { \
+        variableType* ptr = reinterpret_cast<variableType*>(SOFT_LINK_GETPROCADDRESS(framework##Library(), #variableName)); \
+        if (!ptr) \
+            return false; \
+        constant##framework##variableName = *ptr; \
+        return true; \
+    } \
+    bool canLoad_##framework##_##variableName(); \
+    bool canLoad_##framework##_##variableName() \
+    { \
+        static bool loaded = init_##framework##_##variableName(); \
+        return loaded; \
+    } \
+    variableType get_##framework##_##variableName(); \
+    variableType get_##framework##_##variableName() \
+    { \
+        return constant##framework##variableName; \
+    } \
+    }
+
+#define SOFT_LINK_FUNCTION_FOR_HEADER(functionNamespace, framework, functionName, resultType, parameterDeclarations, parameterNames) \
+    namespace functionNamespace { \
+    extern resultType(__cdecl*softLink##framework##functionName) parameterDeclarations; \
+    inline resultType softLink_##framework##_##functionName parameterDeclarations \
+    { \
+        return softLink##framework##functionName parameterNames; \
+    } \
+    }
+
+#define SOFT_LINK_FUNCTION_FOR_SOURCE_WITH_EXPORT(functionNamespace, framework, functionName, resultType, parameterDeclarations, parameterNames, export) \
+    namespace functionNamespace { \
+    static resultType __cdecl init##framework##functionName parameterDeclarations; \
+    export resultType(__cdecl*softLink##framework##functionName) parameterDeclarations = init##framework##functionName; \
+    static resultType __cdecl init##framework##functionName parameterDeclarations \
+    { \
+        softLink##framework##functionName = reinterpret_cast<resultType (__cdecl*)parameterDeclarations>(SOFT_LINK_GETPROCADDRESS(framework##Library(), #functionName)); \
+        RELEASE_ASSERT(softLink##framework##functionName); \
+        return softLink##framework##functionName parameterNames; \
+    } \
+    }
+
+#define SOFT_LINK_FUNCTION_FOR_SOURCE(functionNamespace, framework, functionName, resultType, parameterDeclarations, parameterNames) \
+    SOFT_LINK_FUNCTION_FOR_SOURCE_WITH_EXPORT(functionNamespace, framework, functionName, resultType, parameterDeclarations, parameterNames, )
+
+#define SOFT_LINK_FUNCTION_MAY_FAIL_FOR_HEADER(functionNamespace, framework, functionName, resultType, parameterDeclarations, parameterNames) \
+    WTF_EXTERN_C_BEGIN \
+    resultType functionName parameterDeclarations; \
+    WTF_EXTERN_C_END \
+    namespace functionNamespace { \
+    extern resultType (*softLink##framework##functionName) parameterDeclarations; \
+    bool canLoad_##framework##_##functionName(); \
+    bool init_##framework##_##functionName(); \
+    resultType softLink_##framework##_##functionName parameterDeclarations; \
+    }
+
+#define SOFT_LINK_FUNCTION_MAY_FAIL_FOR_SOURCE(functionNamespace, framework, functionName, resultType, parameterDeclarations, parameterNames) \
+    WTF_EXTERN_C_BEGIN \
+    resultType functionName parameterDeclarations; \
+    WTF_EXTERN_C_END \
+    namespace functionNamespace { \
+    resultType (*softLink##framework##functionName) parameterDeclarations = 0; \
+    bool init_##framework##_##functionName(); \
+    bool init_##framework##_##functionName() \
+    { \
+        ASSERT(!softLink##framework##functionName); \
+        softLink##framework##functionName = reinterpret_cast<resultType (__cdecl*)parameterDeclarations>(SOFT_LINK_GETPROCADDRESS(framework##Library(), #functionName)); \
+        return !!softLink##framework##functionName; \
+    } \
+    \
+    bool canLoad_##framework##_##functionName(); \
+    bool canLoad_##framework##_##functionName() \
+    { \
+        static bool loaded = init_##framework##_##functionName(); \
+        return loaded; \
+    } \
+    \
+    resultType softLink_##framework##_##functionName parameterDeclarations; \
+    resultType softLink_##framework##_##functionName parameterDeclarations \
+    { \
+        ASSERT(softLink##framework##functionName); \
+        return softLink##framework##functionName parameterNames; \
+    } \
+    }
diff --git a/Source/WTF/wtf/win/WTFCRTDebug.h b/Source/WTF/wtf/win/WTFCRTDebug.h
new file mode 100644
index 00000000..0d1d916f
--- /dev/null
+++ b/Source/WTF/wtf/win/WTFCRTDebug.h
@@ -0,0 +1,40 @@
+/*
+ * Copyright (C) 2025 Sony Interactive Entertainment Inc.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE AND ITS CONTRIBUTORS "AS IS" AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL APPLE OR ITS CONTRIBUTORS BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+namespace WTF {
+
+/* MinGW note: upstream declares disableCRTDebugAssertDialog() with
+   WTF_EXPORT_PRIVATE and defines it in wtf/win/WTFCRTDebug.cpp, which is
+   not part of this release tarball. The MSVC CRT debug assert dialog the
+   function disables does not exist on MinGW, so an inline no-op keeps the
+   jsc.cpp call site linkable without adding a new translation unit. */
+#if defined(__MINGW32__)
+inline void disableCRTDebugAssertDialog() { }
+#else
+WTF_EXPORT_PRIVATE void disableCRTDebugAssertDialog();
+#endif
+
+}
diff --git a/Source/cmake/FindGLib.cmake b/Source/cmake/FindGLib.cmake
index b6ff9283..5e6e097d 100644
--- a/Source/cmake/FindGLib.cmake
+++ b/Source/cmake/FindGLib.cmake
@@ -206,7 +206,13 @@ GLib_DeclareComponent(Module  gmodule-2.0  gmodule-2.0)
 GLib_DeclareComponent(Object  gobject-2.0  gobject-2.0)
 GLib_DeclareComponent(Thread  gthread-2.0  gthread-2.0)
 GLib_DeclareComponent(Gio     gio-2.0      gio-2.0 Object)
-GLib_DeclareComponent(GioUnix gio-unix-2.0 gio-2.0 Gio)
+# MinGW: there is no gio-unix-2.0 module; use the gio-2.0 library directly so
+# that code asking for the GioUnix component still links on Windows.
+if (WIN32)
+    GLib_DeclareComponent(GioUnix gio-2.0 gio-2.0 Gio)
+else ()
+    GLib_DeclareComponent(GioUnix gio-unix-2.0 gio-2.0 Gio)
+endif ()
 
 foreach (component IN LISTS GLib_FIND_COMPONENTS)
     GLib_HandleComponent(${component})
diff --git a/Source/cmake/OptionsGTK.cmake b/Source/cmake/OptionsGTK.cmake
index 99c1c6f0..b554a421 100644
--- a/Source/cmake/OptionsGTK.cmake
+++ b/Source/cmake/OptionsGTK.cmake
@@ -97,6 +97,27 @@ else ()
     WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_RESOURCE_USAGE PRIVATE OFF)
 endif ()
 
+if (WIN32)
+    # Features without Windows counterparts (or whose dependencies are not
+    # available in MSYS2/mingw-w64 yet) are disabled by default so that a plain
+    # configure works out of the box. They can still be re-enabled explicitly.
+    WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_JOURNALD_LOG PUBLIC OFF)
+    WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_QUARTZ_TARGET PUBLIC OFF)
+    WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_X11_TARGET PUBLIC OFF)
+    WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_WAYLAND_TARGET PUBLIC OFF)
+    WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_GAMEPAD PUBLIC OFF)
+    WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_SPELLCHECK PUBLIC OFF)
+    WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_SPEECH_SYNTHESIS PUBLIC OFF)
+    WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_REMOTE_INSPECTOR PRIVATE OFF)
+    WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_WEBGL PUBLIC OFF)
+    WEBKIT_OPTION_DEFAULT_PORT_VALUE(USE_GBM PUBLIC OFF)
+    WEBKIT_OPTION_DEFAULT_PORT_VALUE(USE_LIBDRM PUBLIC OFF)
+    WEBKIT_OPTION_DEFAULT_PORT_VALUE(USE_LIBBACKTRACE PUBLIC OFF)
+    WEBKIT_OPTION_DEFAULT_PORT_VALUE(USE_LIBHYPHEN PUBLIC OFF)
+    WEBKIT_OPTION_DEFAULT_PORT_VALUE(USE_LIBSECRET PUBLIC OFF)
+    WEBKIT_OPTION_DEFAULT_PORT_VALUE(USE_VULKAN PUBLIC OFF)
+endif ()
+
 # Public options shared with other WebKit ports. Do not add any options here
 # without approval from a GTK reviewer. There must be strong reason to support
 # changing the value of the option.
@@ -236,6 +257,13 @@ SET_AND_EXPOSE_TO_BUILD(WTF_PLATFORM_WAYLAND ${ENABLE_WAYLAND_TARGET})
 add_definitions(-DBUILDING_GTK__=1)
 add_definitions(-DGETTEXT_PACKAGE="WebKitGTK-${WEBKITGTK_API_VERSION}")
 add_definitions(-DJSC_GLIB_API_ENABLED)
+
+if (WIN32)
+    # MinGW does not ship <sys/mman.h>; the mingw-compat directory provides the
+    # small mmap/munmap/mprotect subset of the API WebKit relies on, plus a
+    # sysconf() declaration. Search it before the default include directories.
+    include_directories(BEFORE "${CMAKE_SOURCE_DIR}/Source/ThirdParty/mingw-compat/include")
+endif ()
 # We do not yet have a systematic way of representing the equivalent of WTF PLATFORM_ macros
 # within Swift. This task is represented within Apple as rdar://168139870.
 # For now, our only immediate need is to determine if we're being built
-- 
2.47.3

